### ISSUE TITLE
[hexagon] Runbook: measure #24 (host logits transport) on any Linux box + S25/S25 Ultra

### ISSUE THREAD

**Issue by dlwlzzero** [dlwlzzero/nntrainer#47]
Self-contained runbook for the #24 device measurement on **any** Linux x86_64 box with a Snapdragon 8 Elite phone (S25 or S25 Ultra) on USB. Nothing here needs the Mac, the container, or the original workstation's `/tmp`. Reading this issue is enough; the result goes into `docs/measurements/24-host-logits.md` on branch `hvx/24-host-logits-v2`.

Phone note: the S25 (non-Ultra) has the same SoC. Record the model in the Notes; if a speed row misses #23's ±5 % band on the S25, write it down as a device/thermal difference, not a failure. Accuracy rows (PPL / top-1 / `E2E gen`) must match regardless of phone.

## 0. One-time machine setup (≈ 30 min + downloads)

| need | how | check |
|---|---|---|
| Hexagon SDK 6.4.x with HEXAGON_Tools 19.0.04 | Qualcomm Software Center or `qpm-cli` (`tools/docker/setup_wizard.sh sdk` shows the qpm commands); any install path | `source <SDK>/setup_sdk_env.source && echo $DEFAULT_TOOLS_VARIANT` → `toolv19` |
| Android NDK r26d | https://dl.google.com/android/repository/android-ndk-r26d-linux.zip, unzip anywhere | `export ANDROID_NDK=<path>` |
| adb | `apt install android-tools-adb` or platform-tools | `adb devices` lists the phone as `device` (USB debugging on) |
| python3 + `transformers>=4.40 safetensors tokenizers sentencepiece numpy` | `pip install ...` (the container's list, tools/docker/Dockerfile:45) | — |
| gcc/g++, md5sum | distro packages | — |

```bash
git clone https://github.com/dlwlzzero/nntrainer && cd nntrainer
git checkout hvx/24-host-logits-v2          # 185a76cb; do not rebase, commit the result on this branch
source <SDK>/setup_sdk_env.source; export ANDROID_NDK=<ndk path>
```

## 1. Model artifacts (skip any file you already have with the same md5)

```bash
# HF checkpoint + tokenizer (public, ~1.5 GB)
python3 -c "from huggingface_hub import snapshot_download; snapshot_download('Qwen/Qwen3-0.6B', local_dir='models/hf', allow_patterns=['*.json','*.txt','model.safetensors'])"
# W8_CX .bin (≈2 min)                       expected md5 7562313bb4cc70410450d0ab3a4fa563, 598,230,528 B
python3 tools/hexagon/make_w8cx_bin.py models/hf models/nntr_qwen3_0.6b_w8cx_DEFAULT.bin
# x86 tools (packer, reference runner)
./tools/hexagon/build_host_x86.sh
# packed images                              qwen3_full: 5abf61bef086368559423daa3bea9a99 / 5926be5703f85531c9c46c1978c25287
./build_x86_hexagon/nntr_hexpack models/nntr_qwen3_0.6b_w8cx_DEFAULT.bin /tmp/qwen3_full
./build_x86_hexagon/nntr_hexpack models/nntr_qwen3_0.6b_w8cx_DEFAULT.bin /tmp/qwen3_full4k --max-seq 4224   # e5c92fad… / b0d8aca9…
# token files: the prompt text is reproduced at the end of docs/measurements/23-sdk64-baseline.md ("Prompt text"); save it as models/eval_23.txt (md5 a838b25d…)
python3 tools/hexagon/make_tokens.py models/hf models/eval_23.txt /tmp/t512.i32 --limit 512      # md5 10bb428f92c792aa7733339f8e6b0da1
# 1024 / 4096 tokens: any English text long enough (speed rows only, no reference needed)
python3 tools/hexagon/make_tokens.py models/hf <long.txt> /tmp/t1024.i32 --limit 1024
python3 tools/hexagon/make_tokens.py models/hf <long.txt> /tmp/t4096.i32 --limit 4096
# your own x86 accuracy reference for this prompt (write the numbers in the Notes; container clang gave 41.2365 / 161, workstation gcc 41.5466 / 164)
./build_x86_hexagon/hexagon_ref_run /tmp/qwen3_full --tokens /tmp/t512.i32 --eval
```
Accuracy anchor on a fresh box: the handoff's table (iii) references the original workstation's P4 prompt (33.0884 / 189). With `t512.i32` generated as above the reference is your `hexagon_ref_run --eval` line instead; the device must land within +0.3 % PPL of it and A vs B must be identical. Say which prompt you used.

## 2. Build (≈ 3 min)

```bash
HEX_ARCH=v75 ./tools/hexagon/build_skel.sh        # build_hexagon/skel/libnntr_htp_skel.so (46.8 KB class)
./tools/hexagon/build_host_test.sh                # prints "remote.h: FASTRPC_MAP_STATIC probe -> ..." — paste into Notes
grep -n "FASTRPC_MAP_STATIC\|FASTRPC_MAP_FD_DELAYED\|remote_register_buf" $HEXAGON_SDK_ROOT/incs/remote.h
grep -n "RPCMEM_DEFAULT_FLAGS\|RPCMEM_FLAG" $HEXAGON_SDK_ROOT/ipc/fastrpc/rpcmem/inc/rpcmem.h
md5sum build_hexagon/skel/libnntr_htp_skel.so build_hexagon/host/hexagon_rpc_test build_hexagon/host/hexagon_e2e_test /tmp/t512.i32 /tmp/t1024.i32 /tmp/t4096.i32
```

## 3. Measure (≈ 25 min; scripts push only files whose md5 changed)

If more than one device is attached, insert the serial after the prefix: `run_e2e_test.sh /tmp/qwen3_full <serial> -- ...` and `run_device_test.sh <serial>`.

```bash
# G1 transport in isolation
./tools/hexagon/run_device_test.sh && python3 tools/hexagon/check_rpc_log.py logs/hexagon/device_test_<stamp>.log
# G2/G3 real step, three host memories
for m in malloc rpcmem static; do ./tools/hexagon/run_e2e_test.sh /tmp/qwen3_full -- --tokens /tmp/t512.i32 --chunk 128 --steps 64 --logits-mem $m; done
# G3 accuracy A vs B (also gives #41 its idle-gap clock number)
for m in malloc rpcmem; do ./tools/hexagon/run_e2e_test.sh /tmp/qwen3_full -- --tokens /tmp/t512.i32 --eval --logits-mem $m; done
# G4 goal rows with the winner (rpcmem, or static if it beat rpcmem by ≥ 0.1 ms in G1)
./tools/hexagon/run_e2e_test.sh /tmp/qwen3_full   -- --tokens /tmp/t1024.i32 --chunk 128 --steps 64 --logits-mem <winner>
./tools/hexagon/run_e2e_test.sh /tmp/qwen3_full4k -- --tokens /tmp/t4096.i32 --chunk 128 --steps 64 --logits-mem <winner>
adb shell md5sum /data/local/tmp/nntr_htp/libnntr_htp_skel.so /data/local/tmp/nntr_htp/hexagon_e2e_test
```
Lines to look for: `RPC_TEST PASS` and `VERDICT: PASS`; per e2e run `E2E init ok …`, `E2E logits_mem <m>`, `E2E gen <64 ids>`, `E2E decode steps=63 median_us=<u> median_pcycles=<c> pcycles_per_us=<r>`, `E2E wall_ms <w>`; `--eval` gives `E2E ppl <x> steps 511 top1 <n> wall_ms <w>` then the same `E2E decode` line. `--logits-mem static` ending in `register_static failed (0x…)` is a valid result: note it, continue without C.

## 4. Hand back

1. Fill tables (i) (ii) (iii) and the Notes in `docs/measurements/24-host-logits.md` (gap ms = median_us/1000 − median_pcycles/2090; decode tok/s = 1e6/median_us). Add the phone model, SDK version and your x86 reference line to the Notes.
2. `git add docs/measurements/24-host-logits.md logs/hexagon/ && git commit -s -m "[docs] Fill the #24 host logits measurement (<phone>, SDK <ver>)" && git push`
3. `gh issue edit 24 -R dlwlzzero/nntrainer --remove-label state:needs-measurement --add-label state:measured`

The next `/hexagon-cycle` reads the file back, decides rpcmem vs static, records #41's opening number and opens the #24 PR. Pass rules are in the handoff's "Results" section; the reviewer applies them, you only fill numbers.

**Comment by dlwlzzero** [comment]
Supervisor 2026-09-18: the #24 measurement this runbook was written for is done (`hvx/24-host-logits-v2` @ `b6d73420`, run on the workstation with the S25 Ultra `R3CY10WM83Y`; results read back on #24). Two things the run taught that a future runbook should carry: name the phone's serial, since the two S25 Ultra units in use differ in cDSP decode clock (1.91 vs 2.05 GHz, HEXAGON.md §7 rule 9); and cite the reference log by the token file it used (`e2e_20260917_111035.log` was the `t512_23.i32` run, not `t512.i32`). Closing as completed.
